api/envoy/extensions/filters/http/cache/v3/cache.proto

```diff
@@ -0,0 +1,68 @@
+syntax = "proto3";
+
+package envoy.extensions.filters.http.cache.v3;
+
+import "envoy/config/route/v3alpha/route_components.proto";
+import "envoy/type/matcher/v3alpha/string.proto";
+
+import "udpa/annotations/versioning.proto";
+
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.extensions.filters.http.cache.v3";
+option java_outer_classname = "CacheProto";
+option java_multiple_files = true;
+
+// [#protodoc-title: HTTP Cache Filter]
+// HTTP Cache Filter :ref:`overview <config_http_filters_cache>`.
+// [#extension: envoy.filters.http.cache]
+message CacheConfig {
+  option (udpa.annotations.versioning).previous_message_type =
+      "envoy.config.filter.http.cache.v2.CacheConfig";
+
+  // Modifies cache key creation by restricting which parts of the URL are included.
+  message KeyCreatorParams {
+    option (udpa.annotations.versioning).previous_message_type =
+        "envoy.config.filter.http.cache.v2.CacheConfig.KeyCreatorParams";
+
+    // If true, exclude the URL scheme from the cache key. Set to true if your origins always
+    // produce the same response for http and https requests.
+    bool exclude_scheme = 1;
+
+    // If true, exclude the host from the cache key. Set to true if your origins' responses don't
+    // ever depend on host.
+    bool exclude_host = 2;
+
+    // If query_parameters_included is nonempty, only query parameters matched
+    // by one or more of its matchers are included in the cache key. Any other
+    // query params will not affect cache lookup.
+    repeated config.route.v3alpha.QueryParameterMatcher query_parameters_included = 3;
+
+    // If query_parameters_excluded is nonempty, query parameters matched by one
+    // or more of its matchers are excluded from the cache key (even if also
+    // matched by query_parameters_included), and will not affect cache lookup.
+    repeated config.route.v3alpha.QueryParameterMatcher query_parameters_excluded = 4;
+  }
+
+  // Name of cache implementation to use.
+  string name = 1 [(validate.rules).string = {min_bytes: 1}];
+
+  // TODO(toddmgreer) implement vary headers
+  //
+  // List of allowed Vary headers. Headers matched by allowed_vary_headers will
+  // be provided to caches in lookups, and the cache filter will only insert
+  // responses whose Vary headers (if any) are matched my a matcher in this
+  // list. If allowed_vary_headers is empty, responses with 'vary' headers will
```

Comment:
I'm probably being dense but after reading this several times I'm not 100% sure what the behavior of Vary is in terms of both lookups and insertion. Can you clarify?

Reply:
Vary is one of the nastier parts of HTTP caching, so I won't be surprised if getting the explanation right takes a few rounds.

The Vary response header tells caches which request headers can affect which response is served. Unfortunately, during a lookup, the cache only learns the value of a response's vary header after computing the cache key and finding a matching response. What follows isn't the only way to deal with this contradiction, but seems to be the best:

1. Insertion:
When inserting a response, all headers listed in the vary header must be recorded. A single request URL may have many different responses for the same URL, and the cache has to store them where they can be found by the URL + the values of the headers referenced by 'vary'. These responses should all have the same 'vary' header. If they don't, we're going to have a bad time.

Example:
(All requests for the same URL)
Request 1 fave-animal header: cat
Response 1 headers: vary: fave-animal
Response 1 body: <cat.png>

Request 2 fave-animal header: dog
Response 2 headers: vary: fave-animal
Response 2 body: <dog.png>

2. Lookup:
When looking up a response, the URL may lead to a set of possible responses. The cache filter gives the cache backend the value of the request's 'fave-animal' header, so the backend  can pick the right response.

There are two problems with the above:
1. How does the cache filter know what headers to send to the cache backend?
2. An unwise (or malicious) backend could set vary headers that make too many requests be separately stored (e.g. varying on user-agent), wasting huge amounts of cache space.

Our solution is allowed-vary-headers:
We whitelist a small set of vary headers we've confirmed don't blow up the cache, perhaps only for trusted upstream servers. If a response's vary header lists anything not whitelisted, we don't cache it. We now know exactly which headers might be needed by the cache backend: those listed in allowed-vary-headers.

Part of the confusion is due to the interaction of cache filter and cache storage implementation:
On insertion, CacheFilter gives request header values to the cache storage implementation, which stores them. On lookup, CacheFilter gives them to the cache storage implementation, which uses them to select the right variant.

Reply:
Thanks this makes sense. Do you mind fleshing out the docs/comments to better cover your explanation? I think it's worth it.

Reply:
I tried to condense it into one description (in cache.proto) with what admins need to know when configuring Envoys, and another (in http_cache.h) for devs extending Envoy caching. Do you think it manages to convey the necessary information?